drafts/FS-1146-scoped-nowarn.md

```diff
@@ -0,0 +1,135 @@
+
+This RFC covers the detailed proposal for this suggestion.
+
+- [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/278)
+- [x] Approved in principle
+- [ ] [Implementation](https://github.com/dotnet/fsharp/pull/17507) (draft / PoC for now)
+- [ ] Discussion (t.b.d.)
+
+# Summary
+
+Allow the #nowarn directive to become scoped (per warning number) by a corresponding #warnon directive.
+
+# Motivation
+
+Today, the #nowarn directive disables the specified warnings for all code from the directive to the end of the file. Usually, however, people want to see all warnings except for some very specific situations. Therefore, a way to disable warnings for some fragment of the code is a much-requested feature.
+
+Quotes form the language suggestion thread:
+
+- "being able to nowarn a few lines is most critical for my workplace"
+- "it would be huge if we could get this"
+- "it would be great to do only small sections that we know we want to ignore rather than the whole file"
+
+Motivating examples, again from the suggestion thread:
+
+```
+// We know f is never called with a None.
+let f (Some a) =    // creates warning 25, which we want to suppress
+    // 2000 loc, where the incomplete match warning is beneficial
+ ```
+
+```
+// We have a union with an obsolete case
+type MyErrors =
+| CoolError1 of CoolError1
+| CoolError2 of CoolError2
+| [<Obsolete>] OldError1 of OldError1
+
+let handler e =
+    match e with
+    | CoolError1 ce -> ...
+    | CoolError2 ce -> ...
+    | OldError ce -> ...   // warning 44, which we want to suppress only here
+```
+
+# Detailed specification
+
+1. The compiler shall recognize a new *compiler directive* (see §12.4 of the F# spec) `#warnon`.
+2. `#warnon` has warning numbers as arguments in the same way as `#nowarn` (including non-string arguments (RFC-1147)).
+3. Currently, `#nowarn` disables warnings from the directive to the end of the file. This will be changed as follows.
+If a certain warning has been disabled by a `#nowarn` directive, then it can be re-enabled by a `#warnon` directive later in the file, carrying the same warning number as an argument. So, the "WarningOff" code checking for that warning number is scoped from the `#nowarn` to the corresponding `#warnon`.
+4. If no corresponding `#warnon` is encountered for a `#nowarn` for a certain warning, the warning is disabled from the `#nowarn` to the end of the file.
+5. Multiple "WarningOff" scopes can be defined in a file by `#nowarn` / `#warnon` pairs.
+6. To use `#warnon` for a warning number without a previous `#nowarn` for that warning number is not allowed (error).
+7. A `#nowarn` directive without arguments is not allowed (error).
+8. `--nowarn` compiler flags are not considered for processing `#warnon` directives. This means that for a warning number `n` that is disabled by a compiler flag
+
+    a) `#warnon` for `n` without previous `#nowarn` for `n` is an error.
+    
```

Comment:
I understand the perspective of protecting against user mistakes, by calling it an error.
But I do not like the UX for the following scenario:
- A huge project, and I want to ignore a warning in just 1 file out of many.

Would the recommendation be to use:
- `--nowarn 123` on project level
- `#nowarn 123` in top of the file
- immediately followed by `#warnon 123` right below it   ?


I am wondering if the repetitive `#nowarn` would be needed.
The F# compilation could be changed to allow the following pairs:
1. Valid pairs of `#nowarn` and `#warnon`
2. A pair of project level `--nowarn` and a file level `#warnon`

Reply:
@T-Gro Thanks for the feedback. This RFC is still in an early phase and therefore marked draft.
My plan was to get a bit further in understanding the implications of these items on the implementation and then open a discussion thread for this RFC.
If it is ok for you I would wait with a response until then.